packages/design-system/src/components/contextMenu/components/subMenuTrigger.js

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import {
+  useDebouncedCallback,
+  useCallback,
+  useEffect,
+  useRef,
+} from '@googleforcreators/react';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { useKeyDownEffect } from '../../keyboard';
+import { FOCUSABLE_SELECTORS, KEYS } from '../../../utils';
+import { useContextMenu } from '../contextMenuProvider';
+import MenuItem from './item';
+
+/**
+ * Extracts all focusable children from an html tree.
+ *
+ * @param {HTMLElement} parent The parent to search
+ * @return {Array.<HTMLElement>} List of focusable elements
+ */
+function getFocusableChildren(parent) {
+  return Array.from(parent.querySelectorAll(FOCUSABLE_SELECTORS.join(', ')));
+}
+
+function SubMenuTrigger({
+  openSubMenu,
+  closeSubMenu,
+  isSubMenuOpen,
+  isRTL = false,
+  subMenuRef,
+  parentMenuRef,
+  ...buttonProps
+}) {
+  const ref = useRef();
+  const pointerTracker = useRef({});
+
+  const { setFocusedId } = useContextMenu(({ actions }) => ({
+    setFocusedId: actions.setFocusedId,
+  }));
+
+  const pointerIsOutside = (node) => {
+    const { x, y, width, height } = node.getBoundingClientRect();
+    const { x: pointerX, y: pointerY } = pointerTracker.current;
+    const buffer = 2;
+    return (
+      pointerX < x - buffer ||
+      pointerX > x + width + buffer ||
+      pointerY < y - buffer ||
+      pointerY > y + height + buffer
+    );
+  };
+
+  const maybeCloseSubMenu = useDebouncedCallback(() => {
+    if (!ref.current || !subMenuRef.current?.firstChild) {
+      return;
+    }
+    // If after 200ms the cursor is not in the submenu and not in itself, leave.
+    if (
+      pointerIsOutside(ref.current) &&
+      pointerIsOutside(subMenuRef.current.firstChild)
```

Comment:
Something here isn't right - or this function isn't correctly invoked at the right times. When you move the pointer out of the submenu quickly, it doesn't close. It only closes if you move it a little bit out. Not exactly sure what's going on here though, it seems a bit inconsistent.

![maybe-close-bug](https://user-images.githubusercontent.com/637548/150609439-4a263612-be8c-4b90-b69c-229d4426450a.gif)

Reply:
Should be fixed now, PTAL.

Reply:
Awesome, works perfect! 🎉